libs/core/include/core/mutex.hpp

```diff
@@ -18,10 +18,93 @@
 //------------------------------------------------------------------------------
 
 #include <mutex>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
 
 namespace fetch {
+class DebugMutex;
 
+class Mutexregister
+{
+public:
+  void RegisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread);
+  void UnregisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread);
+  void FindDeadlock(DebugMutex *mutex, std::thread::id thread);
+
+private:
+  std::mutex                                                     mutex_;
+  std::unordered_map<DebugMutex *, std::thread::id>              lock_owners_;
+  std::unordered_map<std::thread::id, std::vector<DebugMutex *>> acquired_locks_;
+};
+
+void RegisterMutex(DebugMutex *mutex);
+void UnregisterMutex(DebugMutex *mutex);
+void RegisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread);
+void UnregisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread);
+void FindDeadlock(DebugMutex *mutex, std::thread::id thread);
+
+class DebugMutex
+{
+public:
+  DebugMutex()
+    : filename_{"unknown"}
+    , line_{0}
+  {}
+
+  DebugMutex(std::string const &filename, int32_t line)
+    : filename_{filename}
+    , line_{line}
+  {}
+
+  ~DebugMutex() = default;
+
+  void lock()
+  {
+    FindDeadlock(this, std::this_thread::get_id());
+    mutex_.lock();
+    RegisterMutexAcquisition(this, std::this_thread::get_id());
+  }
+
+  void unlock()
+  {
+    UnregisterMutexAcquisition(this, std::this_thread::get_id());
+    mutex_.unlock();
+  }
+
+  bool try_lock()
+  {
+    if (mutex_.try_lock())
+    {
+      RegisterMutexAcquisition(this, std::this_thread::get_id());
+      return true;
+    }
+
+    return false;
+  }
+
+  std::string filename() const
+  {
+    return filename_;
+  }
+
+  int32_t line() const
+  {
+    return line_;
+  };
+
+private:
+  std::mutex  mutex_;
+  std::string filename_{};
+  int32_t     line_{0};
+};
+
+#ifndef NDEBUG
+using Mutex = DebugMutex;
```

Comment:
It is not immediately clear how could the two be possibly interchangeable. `DebugMutex`'s most useful ctor accepts filename and line, while `std::mutex` has none of such. Moreover, wouldn't it be more informative to store the location of lock acquistion, not the mutex construction itself, as is done in `FETCH_LOCK` (`FETCH_LOCK` itself could be modified to make good use of `DebugMutex`).

Reply:
They are not interchangeable in the most general case. `DebugMutex` provides a subset of `std::mutex` capabilities, so as long as the code compiles and runs with `DebugMutex`, it is guaranteed that it will with `std::mutex`